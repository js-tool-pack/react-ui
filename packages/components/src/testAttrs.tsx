import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
