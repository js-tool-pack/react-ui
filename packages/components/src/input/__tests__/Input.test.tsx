import { fireEvent, render } from '@testing-library/react';
import { testAttrs } from '~/testAttrs';
import { InputProps, Input } from '..';

describe('Input', () => {
  testAttrs<InputProps>(({ attrs, ref, ...rest }) => {
    return <Input {...rest} rootAttrs={attrs as any} rootRef={ref as any} />;
  });

  test('basic', () => {
    expect(render(<Input />).container).toMatchSnapshot();
    expect(
      render(<Input attrs={{ className: 'hello' }} />).container,
    ).toMatchSnapshot();
  });

  test('clearable', () => {
    const { container } = render(<Input clearable />);
    expect(container).toMatchSnapshot();

    expect(getInput()).toHaveValue('');
    expect(getClear()).toBeNull();

    fireEvent.change(getInput()!, { target: { value: 'foo bar' } });
    expect(getInput()).toHaveValue('foo bar');
    expect(getClear()).not.toBeNull();

    expect(getInput()).toHaveValue('foo bar');
    fireEvent.click(getClear()!);
    expect(getInput()).toHaveValue('');
  });

  test('size', () => {
    expect(render(<Input />).container.firstChild).toHaveClass('t--size-m');
    expect(render(<Input size="small" />).container.firstChild).toHaveClass(
      't--size-sm',
    );
    expect(render(<Input size="medium" />).container.firstChild).toHaveClass(
      't--size-m',
    );
    expect(render(<Input size="large" />).container.firstChild).toHaveClass(
      't--size-lg',
    );
  });

  test('prefix', () => {
    expect(
      render(<Input prefix="foo" />).container.firstChild,
    ).toMatchSnapshot();
  });

  test('suffix', () => {
    expect(
      render(<Input suffix="bar" />).container.firstChild,
    ).toMatchSnapshot();
  });

  test('loading', () => {
    expect(render(<Input loading />).container.firstChild).toMatchSnapshot();
  });

  test('textarea', () => {
    expect(
      render(<Input type="textarea" />).container.firstChild,
    ).toMatchSnapshot();
  });

  test('count', () => {
    expect(render(<Input showCount />).container.firstChild).toMatchSnapshot();
    expect(
      render(<Input maxLength={20} showCount />).container.firstChild,
    ).toMatchSnapshot();
    expect(
      render(<Input type="textarea" showCount />).container.firstChild,
    ).toMatchSnapshot();
    expect(
      render(<Input type="textarea" maxLength={20} showCount />).container
        .firstChild,
    ).toMatchSnapshot();

    expect(
      render(
        <Input value="👨‍👨‍👦‍👦" maxLength={20} showCount />,
      ).container.querySelector('.t-input__count'),
    ).toHaveTextContent('11 / 20');

    const segmenter = new Intl.Segmenter('fr', {
      granularity: 'grapheme',
    });
    expect(
      render(
        <Input
          count={(value) => Array.from(segmenter.segment(value)).length}
          value="👨‍👨‍👦‍👦"
          maxLength={20}
          showCount
        />,
      ).container.querySelector('.t-input__count'),
    ).toHaveTextContent('1 / 20');
  });

  describe('password', () => {
    test('basic', () => {
      const { container } = render(<Input type="password" />);
      expect(container.firstChild).toMatchSnapshot();
      expect(getInput()).toHaveValue('');
    });

    test('value', () => {
      render(<Input type="password" value="123" />);
      expect(getInput()).not.toHaveAttribute('value');

      fireEvent.focus(getInput()!);
      expect(getInput()).not.toHaveAttribute('value');

      fireEvent.blur(getInput()!);
      expect(getInput()).not.toHaveAttribute('value');

      fireEvent.change(getInput()!, { target: { value: '321' } });
      expect(getInput()).not.toHaveAttribute('value');
    });

    test('showPasswordOn = click', () => {
      render(<Input showPasswordOn="click" type="password" value="123" />);

      expect(getInput()).toHaveValue('123');
      expect(getInput()).not.toHaveAttribute('value');

      fireEvent.click(getSwitch()!);
      expect(getInput()).toHaveAttribute('value');

      fireEvent.click(getSwitch()!);
      expect(getInput()).not.toHaveAttribute('value');
    });

    test('showPasswordOn = mouseDown', () => {
      render(<Input showPasswordOn="mouseDown" type="password" value="123" />);

      expect(getInput()).toHaveValue('123');
      expect(getInput()).not.toHaveAttribute('value');

      fireEvent.mouseDown(getSwitch()!);
      expect(getInput()).toHaveAttribute('value');

      fireEvent.mouseUp(getSwitch()!);
      expect(getInput()).not.toHaveAttribute('value');
    });

    function getSwitch() {
      return document.querySelector<HTMLElement>('.t-input__switch');
    }
  });

  function getClear(container?: HTMLElement) {
    return (container || document).querySelector<HTMLElement>(
      '.t-input__clear',
    );
  }
  function getInput(container?: HTMLElement) {
    return (container || document).querySelector<HTMLInputElement>('input');
  }
});
