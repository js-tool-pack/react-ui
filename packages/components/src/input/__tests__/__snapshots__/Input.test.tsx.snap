// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input basic 1`] = `
<div>
  <label
    class="t-input t--size-m"
  >
    <input
      type="text"
      value=""
    />
  </label>
</div>
`;

exports[`Input basic 2`] = `
<div>
  <label
    class="t-input t--size-m"
  >
    <input
      class="hello"
      type="text"
      value=""
    />
  </label>
</div>
`;

exports[`Input clearable 1`] = `
<div>
  <label
    class="t-input t--size-m t-input--clearable"
  >
    <input
      type="text"
      value=""
    />
  </label>
</div>
`;

exports[`Input count 1`] = `
<label
  class="t-input t--size-m"
>
  <input
    type="text"
    value=""
  />
  <div
    class="t-input__suffix"
  >
    <div
      class="t-input__count"
    >
      <span>
        0
      </span>
    </div>
  </div>
</label>
`;

exports[`Input count 2`] = `
<label
  class="t-input t--size-m"
>
  <input
    type="text"
    value=""
  />
  <div
    class="t-input__suffix"
  >
    <div
      class="t-input__count"
    >
      <span>
        0
         / 
        20
      </span>
    </div>
  </div>
</label>
`;

exports[`Input count 3`] = `
<label
  class="t-input t--size-m t-input--textarea"
>
  <textarea
    rows="3"
  />
  <div
    class="t-input__count"
  >
    <span>
      0
    </span>
  </div>
</label>
`;

exports[`Input count 4`] = `
<label
  class="t-input t--size-m t-input--textarea"
>
  <textarea
    rows="3"
  />
  <div
    class="t-input__count"
  >
    <span>
      0
       / 
      20
    </span>
  </div>
</label>
`;

exports[`Input loading 1`] = `
<label
  class="t-input t--size-m t-input--loading"
>
  <input
    type="text"
    value=""
  />
  <div
    class="t-input__suffix"
  >
    <i
      class="t-icon t-input__icon t-input__loading"
      role="img"
    >
      <svg
        class="t-loading-icon"
        viewBox="0 0 66 66"
      >
        <circle
          class="t-loading-icon__path"
          cx="33"
          cy="33"
          fill="none"
          r="30"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-width="6"
        />
      </svg>
    </i>
  </div>
</label>
`;

exports[`Input password basic 1`] = `
<label
  class="t-input t--size-m"
>
  <input
    type="password"
  />
  <div
    class="t-input__suffix"
  >
    <i
      class="t-icon t-input__icon t-input__switch"
      role="img"
    >
      <svg
        viewBox="0 0 1102 1024"
      >
        <path
          d="M1080.950154 498.54358967l-0.078769-0.157538-0.157539-0.07877c-44.740923-94.444308-98.461538-170.692923-160.846769-228.903384l-62.385231 62.464c53.799385 49.467077 100.273231 115.318154 140.445539 198.025846-102.557538 212.125538-247.886769 312.713846-446.54277 312.713846-63.172923 0-120.989538-10.161231-173.764923-30.798769l-67.347692 67.268923c71.995077 34.658462 152.260923 52.224 241.112615 52.224 236.544 0 412.750769-123.667692 529.48677-369.585231a74.043077 74.043077 0 0 0 0.078769-63.172923z m-78.769231-408.812308l-52.145231-52.224a9.846154 9.846154 0 0 0-13.942154 0L792.418462 181.10358967C720.502154 146.28758967 640.393846 128.80082067 551.384615 128.80082067c-236.544 0-412.750769 123.667692-529.486769 369.585231V498.54358967a74.200615 74.200615 0 0 0 0 63.409231c44.819692 94.365538 98.461538 170.692923 160.846769 228.982154l-123.904 123.667692a9.846154 9.846154 0 0 0 0 13.942154l52.145231 52.224a9.846154 9.846154 0 0 0 13.942154 0L1002.180923 103.43712867a9.767385 9.767385 0 0 0 0-13.784616z m-589.981538 471.591385a137.846154 137.846154 0 0 1 165.572923-165.572923L412.120615 561.32266667z m227.091692-227.091692a216.772923 216.772923 0 0 0-288.610462 288.610461L245.287385 728.23466667c-53.799385-49.545846-100.273231-115.318154-140.445539-198.025846C207.478154 318.08328167 352.886154 217.49497467 551.384615 217.49497467c63.172923 0 120.989538 10.161231 173.764923 30.798769l-85.858461 85.937231zM546.500923 667.89743567c-7.876923 0-15.675077-0.630154-23.158154-1.969231l-62.936615 62.857847a216.615385 216.615385 0 0 0 284.908308-284.908308l-62.857847 62.936615A138.003692 138.003692 0 0 1 546.422154 667.89743567z"
          fill="currentColor"
        />
      </svg>
    </i>
  </div>
</label>
`;

exports[`Input password basic 2`] = `
<label
  class="t-input t--size-m"
>
  <input
    type="password"
  />
  <div
    class="t-input__suffix"
  >
    <i
      class="t-icon t-input__icon t-input__switch"
      role="img"
    >
      <svg
        viewBox="0 0 1102 1024"
      >
        <path
          d="M1080.950154 498.54358967l-0.078769-0.157538-0.157539-0.07877c-44.740923-94.444308-98.461538-170.692923-160.846769-228.903384l-62.385231 62.464c53.799385 49.467077 100.273231 115.318154 140.445539 198.025846-102.557538 212.125538-247.886769 312.713846-446.54277 312.713846-63.172923 0-120.989538-10.161231-173.764923-30.798769l-67.347692 67.268923c71.995077 34.658462 152.260923 52.224 241.112615 52.224 236.544 0 412.750769-123.667692 529.48677-369.585231a74.043077 74.043077 0 0 0 0.078769-63.172923z m-78.769231-408.812308l-52.145231-52.224a9.846154 9.846154 0 0 0-13.942154 0L792.418462 181.10358967C720.502154 146.28758967 640.393846 128.80082067 551.384615 128.80082067c-236.544 0-412.750769 123.667692-529.486769 369.585231V498.54358967a74.200615 74.200615 0 0 0 0 63.409231c44.819692 94.365538 98.461538 170.692923 160.846769 228.982154l-123.904 123.667692a9.846154 9.846154 0 0 0 0 13.942154l52.145231 52.224a9.846154 9.846154 0 0 0 13.942154 0L1002.180923 103.43712867a9.767385 9.767385 0 0 0 0-13.784616z m-589.981538 471.591385a137.846154 137.846154 0 0 1 165.572923-165.572923L412.120615 561.32266667z m227.091692-227.091692a216.772923 216.772923 0 0 0-288.610462 288.610461L245.287385 728.23466667c-53.799385-49.545846-100.273231-115.318154-140.445539-198.025846C207.478154 318.08328167 352.886154 217.49497467 551.384615 217.49497467c63.172923 0 120.989538 10.161231 173.764923 30.798769l-85.858461 85.937231zM546.500923 667.89743567c-7.876923 0-15.675077-0.630154-23.158154-1.969231l-62.936615 62.857847a216.615385 216.615385 0 0 0 284.908308-284.908308l-62.857847 62.936615A138.003692 138.003692 0 0 1 546.422154 667.89743567z"
          fill="currentColor"
        />
      </svg>
    </i>
  </div>
</label>
`;

exports[`Input prefix 1`] = `
<label
  class="t-input t--size-m"
>
  <div
    class="t-input__prefix"
  >
    foo
  </div>
  <input
    type="text"
    value=""
  />
</label>
`;

exports[`Input suffix 1`] = `
<label
  class="t-input t--size-m"
>
  <input
    type="text"
    value=""
  />
  <div
    class="t-input__suffix"
  >
    bar
  </div>
</label>
`;

exports[`Input textarea 1`] = `
<label
  class="t-input t--size-m t-input--textarea"
>
  <textarea
    rows="3"
  />
</label>
`;
